Trimmomatic

pjbriggs opened the pull request:
FOR CONTRIBUTOR:
* [x] - I have read the [CONTRIBUTING.md](https://github.com/galaxyproject/tools-iuc/blob/master/CONTRIBUTING.md) document and this tool is appropriate for the tools-iuc repo.
* [x] - License permits unrestricted use (educational + commercial)
* [x] - This PR adds a new tool or tool collection
* [ ] - This PR updates an existing tool or tool collection
* [ ] - This PR does something else (explain below)

PR imports the `trimmomatic` tool from the repo https://github.com/fls-bioinformatics-core/galaxy-tools with the intention of contributing the tool to the IUC (after discussion with @bernt-matthias). 

The toolshed repository owner has been kept as `pjbriggs` to preserve the tool lineage; the `iuc@galaxyproject` user has been added as an administrator for the toolshed repo (the `Intergalatic Utilities Commission` group is now also an admin group).

The following external PRs and issues may be relevant for background:

* https://github.com/fls-bioinformatics-core/galaxy-tools/pull/90
* https://github.com/fls-bioinformatics-core/galaxy-tools/issues/91
* https://github.com/fls-bioinformatics-core/galaxy-tools/pull/92

bernt-matthias commented:
My initial intention here was to get a container for the tool. But I think it's great to have it in IUC anyway.

bernt-matthias commented on tools/trimmomatic/trimmomatic.xml:
```diff
@@ -0,0 +1,531 @@
+  #if $quality_score
+    $quality_score
+  #end if
+  2>&1 | tee trimmomatic.log &&
```
This redirects `stderr` and `stdout`. Are both needed, or would `stdout` be sufficient?

bernt-matthias reviewed (commented):
The linter "fails" with a warning

```
 .. WARNING: Test should specify 'expect_num_outputs' if outputs have filters
```

if you like you may add this.

mvdbeek commented:
I'm a little sad about this, I think having tools in external repos is great, and not everything needs to follow the IUC standards closely (I particularly disagree that there is any requirement whatsoever for including a repository in planemo-monitor). If you'd like less maintenance burden then that's great of course, but you've been doing amazing so far :).

pjbriggs commented:
@bernt-matthias re the linting issue (`expect_num_outputs`), I will take a look and try to address - thanks.

bernt-matthias commented:
Adding yourself for the tool here https://github.com/galaxyproject/tools-iuc/blob/main/.github/CODEOWNERS would definitely be a good idea.

bernt-matthias commented on tools/trimmomatic/trimmomatic.xml:
```diff
@@ -0,0 +1,531 @@
+	See similar fix for snpSift
+	https://github.com/galaxyproject/tools-iuc/commit/b5e2080a7afdea9fa476895693b6115824c6fbb9
+    -->
+    <requirement type="package" version="8.25">coreutils</requirement>
```
Seems that conda can't resolve this anymore. It should work with a more recent version of coreutils:

```suggestion
    <requirement type="package" version="9.4">coreutils</requirement>
```

pjbriggs replied:
Thanks - updated!

bernt-matthias commented:
> Adding yourself for the tool here https://github.com/galaxyproject/tools-iuc/blob/main/.github/CODEOWNERS would definitely be a good idea.

Forgot that unfortunately only IUC members can be listed in codeowners. 

Test 1 and 14 is failing due to differences in the `err_file`


```diff
--- local_file
+++ history_data
@@ -1,5 +1,5 @@
 TrimmomaticSE: Started with arguments:
- -threads 1 fastq_in.fastqsanger fastq_out.fastqsanger SLIDINGWINDOW:4:20 -trimlog trimlog
+ -threads 4 fastq_in.fastqsanger fastq_out.fastqsanger SLIDINGWINDOW:4:20 -trimlog trimlog
 Quality encoding detected as phred33
 Input Reads: 10 Surviving: 8 (80.00%) Dropped: 2 (20.00%)
 TrimmomaticSE: Completed successfully
```

Reason is that tool tests run here with the number of cores available in the action runner (which may change over time). You could use `compare="re_match"` and relax the number of threads. But you would need to format the content of the test file as regular expression, e.g.:

```
TrimmomaticSE: Started with arguments:
threads [0-9]+ fastq_in\.fastqsanger fastq_out\.fastqsanger SLIDINGWINDOW:4:20 -trimlog trimlog
Quality encoding detected as phred33
Input Reads: 10 Surviving: 8 \(80\.00%\) Dropped: 2 \(20\.00%\)
TrimmomaticSE: Completed successfully
```

pjbriggs commented:
> Adding yourself for the tool here https://github.com/galaxyproject/tools-iuc/blob/main/.github/CODEOWNERS would definitely be a good idea.

Thanks - also done.

pjbriggs commented:
@bernt-matthias I've updated the tool tests to use `re_match` when checking the `err_file` output (thanks for identifying the issue and suggesting the fix), and also fixed the final linting issue with the `do_illuminaclip` conditional being a boolean (by converting to `select`).

Are you happy to mark the conversation about redirecting both `stderr` and `stdout` as resolved?

bernt-matthias commented on .github/CODEOWNERS:
```diff
@@ -28,4 +28,5 @@
 /tools/seurat/                          @nsoranzo
 /tools/sickle/                          @nsoranzo
 /tools/stringtie/                       @nsoranzo
+/tools/trimmomatic/                     @pjbriggs
```
```suggestion
```

github tells me that we are not allowed to do this (only persons with write access can be listed here) .. unfortunatelly. 

Guess it would be cool to have a better mechanism here.

bernt-matthias replied:
Si, I'm absolutely fine with keeping it as is. But I'm not sure about side effects.

pjbriggs replied:
No problem - I've removed the line now. (I thought it might be okay because I seem to be a member of `galaxyproject`, but I guess that wasn't sufficient. Hopefully github is happy again! :)

bgruening replied:
I see only one solution here ... Peter needs to become an IUC member :)

pjbriggs commented:
Thanks @bernt-matthias and @bgruening!
